File: src/winphone/MPin/MPinSDK/InterfaceImplementations/HTTPConnector.cs
// Copyright (c) 2012-2015, Certivox
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
// For full details regarding our CertiVox terms of service please refer to
// the following links:
//  * Our Terms and Conditions -
//    http://www.certivox.com/about-certivox/terms-and-conditions/
//  * Our Security and Privacy -
//    http://www.certivox.com/about-certivox/security-privacy/
//  * Our Statement of Position and Our Promise on Software Patents -
//    http://www.certivox.com/about-certivox/patents/

using MPinRC;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using System.Reflection;
using System.Threading;
using Windows.Web.Http;
using Windows.Web.Http.Headers;
using Windows.Storage.Streams;
using Windows.Security.Cryptography;
using Windows.Foundation;
using System.Runtime.InteropServices.WindowsRuntime;

namespace MPinSDK
{
    /// <summary>
    /// A class implementing the <see cref="T:MPinRC.IHttpRequest"/> interface that the Core uses to make HTTP requests.
    /// </summary>
    class HTTPConnector : IHttpRequest
    {
        #region Members
        public const int DEFAULT_TIMEOUT = 30 * 1000;

        private IDictionary<String, String> requestHeaders = new Dictionary<string, string>();
        private IDictionary<String, String> queryParams = new Dictionary<string, string>();
        private String requestBody;
        private int timeout = DEFAULT_TIMEOUT;
        private String errorMessage;
        private int statusCode;
        private IDictionary<String, String> responseHeaders = new Dictionary<string, string>();
        private String responseData;
        #endregion

        #region C'tor
        /// <summary>
        /// Initializes a new instance of the <see cref="HTTPConnector"/> class.
        /// </summary>
        public HTTPConnector()
            : base()
        { }
        #endregion 
        
        #region IHTTPRequest
        /// <summary>
        /// Sets the headers for the HTTP Request.
        /// </summary>
        /// <param name="headers">The headers key/value map to be set.</param>
        public void SetHeaders(IDictionary<string, string> headers)
        {
            this.requestHeaders = headers;
        }

        public void SetQueryParams(IDictionary<string, string> queryParams)
        {
            this.queryParams = queryParams;
        }

        public void SetContent(string data)
        {
            this.requestBody = data;
        }

        public void SetTimeout(int seconds)
        {
            if (seconds <= 0) throw new ArgumentException("The timeout could not be set to a negative value!");
            this.timeout = seconds;
        }

        public bool Execute(Windows.Web.Http.HttpMethod method, string url)
        {
            bool result = false;
            Task.Run( async () =>
                {
                    result =  await ExecuteAsync(method, url);
                }).Wait();

            return result;            
        }
    
        public string GetExecuteErrorMessage()
        {
            return errorMessage;
        }

        public int GetHttpStatusCode()
        {
            return statusCode;
        }

        public IDictionary<string, string> GetResponseHeaders()
        {
            return responseHeaders;
        }

        public string GetResponseData()
        {
            return responseData;
        }
        #endregion // IHTTPRequest

        #region Methods
        
        private async Task<bool> ExecuteAsync(Windows.Web.Http.HttpMethod method, string url)
        {
            ClearResponseData();

            if (string.IsNullOrEmpty(url)) 
                throw new ArgumentException("The url value is empty!");

            String fullUrl = url;
            if (queryParams != null && queryParams.Count > 0)
            {
                ICollection<String> keyEnum = queryParams.Keys;
                fullUrl += "?";
                foreach (var key in keyEnum)
                {
                    fullUrl = key + "=" + queryParams[key] + "&";
                }

                fullUrl = fullUrl.Substring(0, fullUrl.Length - 1);
            }

            bool successful = true;
            try
            {
                await this.SendRequest(fullUrl, method, requestBody, requestHeaders);
            }
            catch (Exception e)
            {
                errorMessage = e.Message;
                successful = false;
            }
            finally
            {
                ClearRequestData();
            }

            return successful;
        }

        private void ClearRequestData()
        {
            this.requestHeaders.Clear();
            this.queryParams.Clear();
            this.requestBody = string.Empty;
            this.timeout = 0;
        }

        private void ClearResponseData()
        {
            this.statusCode = 0;
            this.responseHeaders.Clear();
            this.responseData = string.Empty;
            this.errorMessage = string.Empty;
        }
       
        protected async Task SendRequest(String serviceURL, Windows.Web.Http.HttpMethod http_method, String requestBody, IDictionary<String, String> requestProperties)
        {
            HttpClient httpClient = new HttpClient();
            CancellationTokenSource cts = new CancellationTokenSource();
            try
            {
                Uri resourceAddress = new Uri(serviceURL);
                HttpRequestMessage request = new HttpRequestMessage(http_method, resourceAddress);

                if (!string.IsNullOrEmpty(requestBody))
                {
                    Stream stream = new MemoryStream(Encoding.UTF8.GetBytes(requestBody ?? ""));
                    request.Content = new HttpStreamContent(stream.AsInputStream());
                }

                if (requestProperties != null && requestProperties.Count > 0)
                {
                    foreach (var key in requestProperties.Keys)
                    {
                        request.Properties.Add(new KeyValuePair<string, object>(key, requestProperties[key]));                        
                    }
                }
                
                HttpResponseMessage response = await httpClient.SendRequestAsync(
                    request,
                    HttpCompletionOption.ResponseHeadersRead).AsTask(cts.Token);

                SetResponseHeaders(response.Headers);
                this.responseData = await response.Content.ReadAsStringAsync();
                this.statusCode = (int)response.StatusCode;

            }
            catch (TaskCanceledException tce)
            {
                this.responseData = string.Empty;
                this.errorMessage = "Request canceled!";
                throw tce;
            }
            catch (Exception ex)
            {
                this.responseData = string.Empty;
                this.errorMessage = "Error: " + ex.Message;
                throw ex;
            }
        }
       
        private string ToString(StreamReader sr)
        {
            try
            {
                int i = 0, length = 512;
                char[] buf = new char[length];
                StringBuilder str = new StringBuilder();

                while ((i = sr.Read(buf, 0, length)) != -1)
                {
                    str.Append(buf, 0, i);
                }

                return str.ToString();
            }
            finally
            {
                if (sr != null)
                {
                    sr.Dispose();
                }
            }
        }
        
        private void SetResponseHeaders(HttpResponseHeaderCollection headers)
        {
            if (responseHeaders == null)
            {
                responseHeaders = new Dictionary<String, String>();
            }
            else
            {
                responseHeaders.Clear();
            }

            foreach (var key in headers.Keys)
            {
                String properties = "";
                foreach (var s in headers[key])
                {
                    properties += s;
                }
                responseHeaders.Add(new KeyValuePair<string, string>(key, properties));
            }
        }
        #endregion // Methods

        #region HTTPErrorException
        public class HTTPErrorException : Exception
        {
            public int StatusCode
            {
                get;
                set;
            }

            public HTTPErrorException()
                : base()
            { }

            public HTTPErrorException(String message)
                : base(message)
            { }

            public HTTPErrorException(String message, int statusCode)
                : base(message)
            {
                this.StatusCode = statusCode;
            }
        }
        #endregion // HTTPErrorException
    }
}

File: src/android/MPinLib/src/com/certivox/net/HTTPConnector.java
/*******************************************************************************
 * Copyright (c) 2012-2015, Certivox All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
 * following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
 * disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
 * following disclaimer in the documentation and/or other materials provided with the distribution.
 * 
 * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
 * products derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * For full details regarding our CertiVox terms of service please refer to the following links:
 * 
 * * Our Terms and Conditions - http://www.certivox.com/about-certivox/terms-and-conditions/
 * 
 * * Our Security and Privacy - http://www.certivox.com/about-certivox/security-privacy/
 * 
 * * Our Statement of Position and Our Promise on Software Patents - http://www.certivox.com/about-certivox/patents/
 ******************************************************************************/
package com.certivox.net;


import java.io.DataOutputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Enumeration;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;

import android.net.Uri;
import android.text.TextUtils;


public class HTTPConnector implements IHTTPRequest {

    private Hashtable<String, String> requestHeaders;
    private Hashtable<String, String> queryParams;
    private String                    requestBody;
    private int                       timeout = DEFAULT_TIMEOUT;
    private String                    errorMessage;
    private int                       statusCode;
    private Hashtable<String, String> responseHeaders;
    private String                    responseData;

    private final static String OS_CLASS_HEADER = "X-MIRACL-OS-Class";
    private final static String OS_CLASS_VALUE = "android";

    // / only for test !!!!
    public String getContent() {

        return requestBody;
    }


    public Hashtable<String, String> RequestHeaders() {
        return requestHeaders;
    }


    public HTTPConnector() {
        super();
    }


    protected HttpURLConnection getConnection(String serviceURL, boolean output)
            throws MalformedURLException, IOException {
        if (serviceURL.startsWith("/")) {
            serviceURL = "http://ec2-54-77-232-113.eu-west-1.compute.amazonaws.com" + serviceURL;
        }

        HttpURLConnection httpConnection = (HttpURLConnection) new URL(serviceURL).openConnection();
        httpConnection.setDoInput(true);
        httpConnection.setDoOutput(output);
        return httpConnection;
    }


    protected String HttpMethodMapper(int method) {
        switch (method) {
        case GET:
            return HTTP_GET;
        case POST:
            return HTTP_POST;
        case PUT:
            return HTTP_PUT;
        case DELETE:
            return HTTP_DELETE;
        case OPTIONS:
            return HTTP_OPTIONS;
        default:
            return HTTP_PATCH;

        }
    }


    protected String sendRequest(String serviceURL, String http_method, String requestBody,
            Hashtable<String, String> requestProperties) throws IOException, HTTPErrorException {

        HttpURLConnection connection = null;
        DataOutputStream dos = null;
        String response = "200 OK";

        try {
            connection = getConnection(serviceURL, !TextUtils.isEmpty(requestBody));

            connection.setRequestMethod(http_method);
            connection.setConnectTimeout(timeout);

            if (requestProperties != null) {
                if (!requestProperties.isEmpty()) {
                    Enumeration<String> keyEnum = requestProperties.keys();
                    while (keyEnum.hasMoreElements()) {
                        String key = keyEnum.nextElement();
                        connection.setRequestProperty(key, requestProperties.get(key));
                    }
                }
            }
            
            connection.setRequestProperty(OS_CLASS_HEADER, OS_CLASS_VALUE);

            if (!TextUtils.isEmpty(requestBody)) {
                dos = new DataOutputStream(connection.getOutputStream());
                dos.writeBytes(requestBody);
            }

            // Starts the query
            connection.connect();

            try {
                statusCode = connection.getResponseCode();
            } catch (IOException e) {
                statusCode = connection.getResponseCode();
                if (statusCode != 401) {
                    throw e;
                }
            }

            responseHeaders = new Hashtable<String, String>();
            Map<String, List<String>> map = connection.getHeaderFields();
            for (Map.Entry<String, List<String>> entry : map.entrySet()) {
                List<String> propertyList = entry.getValue();
                String properties = "";
                for (String s : propertyList) {
                    properties += s;
                }
                String key = entry.getKey();
                if (key == null)
                    continue;
                responseHeaders.put(entry.getKey(), properties);
            }
            response = toString(connection.getInputStream());

        } finally {
            if (dos != null) {
                dos.close();
            }
            if (connection != null) {
                connection.disconnect();
            }
        }

        return response;
    }


    protected String sendRequest(String serviceURL, String http_method, String requestBody)
            throws IOException, HTTPErrorException {
        return sendRequest(serviceURL, http_method, requestBody, null);
    }


    protected String sendRequest(String serviceURL, String http_method) throws IOException, HTTPErrorException {
        return sendRequest(serviceURL, http_method, null);
    }


    @Override
    public void SetHeaders(Hashtable<String, String> headers) {
        this.requestHeaders = headers;
    }


    @Override
    public void SetQueryParams(Hashtable<String, String> queryParams) {
        this.queryParams = queryParams;
    }


    @Override
    public void SetContent(String data) {
        this.requestBody = data;
    }


    @Override
    public void SetTimeout(int seconds) {
        if (seconds <= 0)
            throw new IllegalArgumentException();
        this.timeout = seconds;
    }


    @Override
    public boolean Execute(int method, String url) {
        if (TextUtils.isEmpty(url))
            throw new IllegalArgumentException();

        String fullUrl = url;
        if (queryParams != null) {
            if (!queryParams.isEmpty()) {
                Enumeration<String> keyEnum = queryParams.keys();
                fullUrl += "?";
                while (keyEnum.hasMoreElements()) {
                    String key = keyEnum.nextElement();
                    fullUrl = key + "=" + queryParams.get(key) + "&";
                }
                fullUrl = fullUrl.substring(0, fullUrl.length() - 1);
            }
        }

        // TODO temporary hack
        Uri uri = Uri.parse(fullUrl);
        if ("wss".equals(uri.getScheme()))
            fullUrl = uri.buildUpon().scheme("https").build().toString();

        try {
            responseData = sendRequest(fullUrl, HttpMethodMapper(method), requestBody, requestHeaders);
        } catch (FileNotFoundException e) {
            // No data in response
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
            errorMessage = e.getLocalizedMessage();
            return false;
        }

        return true;
    }


    @Override
    public String GetExecuteErrorMessage() {
        return errorMessage;
    }


    @Override
    public int GetHttpStatusCode() {
        return statusCode;
    }


    @Override
    public Hashtable<String, String> GetResponseHeaders() {
        return responseHeaders;
    }


    @Override
    public String GetResponseData() {
        return responseData;
    }

    @SuppressWarnings("serial")
    public class HTTPErrorException extends Exception {

        private int statusCode;


        public HTTPErrorException() {
            // TODO Auto-generated constructor stub
        }


        public HTTPErrorException(String message) {
            super(message);
            // TODO Auto-generated constructor stub
        }


        public HTTPErrorException(String message, int statusCode) {
            super(message);
            setStatusCode(statusCode);
        }


        public HTTPErrorException(Throwable cause) {
            super(cause);
            // TODO Auto-generated constructor stub
        }


        public HTTPErrorException(String message, Throwable cause) {
            super(message, cause);
            // TODO Auto-generated constructor stub
        }


        public int getStatusCode() {
            return statusCode;
        }


        public void setStatusCode(int statusCode) {
            this.statusCode = statusCode;
        }
    }


    private static String toString(InputStream is) throws IOException {
        InputStreamReader isr = null;
        try {
            isr = new InputStreamReader(is, "UTF-8");
            char[] buf = new char[512];
            StringBuilder str = new StringBuilder();
            int i = 0;
            while ((i = isr.read(buf)) != -1)
                str.append(buf, 0, i);
            return str.toString();
        } finally {
            if (isr != null) {
                isr.close();
            }
        }
    }

}
